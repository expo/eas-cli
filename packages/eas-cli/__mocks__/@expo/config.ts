const getConfig = jest.fn();

export { getConfig };
