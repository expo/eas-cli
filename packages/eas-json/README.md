# @expo/eas-json

A library for interacting with the eas.json
