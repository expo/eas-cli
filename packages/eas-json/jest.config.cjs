module.exports = {
  ...require('../../jest/jest.config.cjs'),
  rootDir: __dirname,
  roots: ['src', '__mocks__'],
};
