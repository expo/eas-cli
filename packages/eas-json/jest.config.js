import { dirname } from 'path';
import { fileURLToPath } from 'url';

import baseConfig from '../../jest/jest.config.js';

const filename = fileURLToPath(import.meta.url);
// eslint-disable-next-line no-underscore-dangle
const __dirname = dirname(filename);

export default {
  ...baseConfig,
  rootDir: __dirname,
  roots: ['src', '__mocks__'],
};
