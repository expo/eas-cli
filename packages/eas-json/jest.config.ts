import type { Config } from 'jest';

// eslint-disable-next-line import/no-relative-packages
import baseConfig from '../../jest/jest.config';

const config: Config = {
  ...baseConfig,
  rootDir: __dirname,
  roots: ['src', '__mocks__'],
};

export default config;
