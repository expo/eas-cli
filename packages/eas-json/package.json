{
  "name": "@expo/eas-json",
  "description": "A library for interacting with the eas.json",
  "version": "0.33.0",
  "author": "Expo <support@expo.dev>",
  "bugs": "https://github.com/expo/eas-cli/issues",
  "dependencies": {
    "@expo/eas-build-job": "0.2.55",
    "@expo/json-file": "8.2.33",
    "chalk": "4.1.2",
    "env-string": "1.0.1",
    "fs-extra": "10.0.0",
    "joi": "17.4.2",
    "log-symbols": "4.1.0",
    "tslib": "2.3.1"
  },
  "devDependencies": {
    "@types/fs-extra": "9.0.12",
    "memfs": "3.2.2"
  },
  "engines": {
    "node": ">=12.0.0"
  },
  "homepage": "https://github.com/expo/eas-cli",
  "license": "MIT",
  "main": "build/index.js",
  "repository": "expo/eas-cli",
  "scripts": {
    "build": "tsc --project tsconfig.build.json",
    "watch": "yarn build --watch --preserveWatchOutput",
    "typecheck": "tsc",
    "prepack": "rm -rf build && yarn build",
    "test": "jest"
  },
  "publishConfig": {
    "access": "public"
  }
}
