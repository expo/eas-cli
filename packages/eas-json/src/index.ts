export { AndroidReleaseStatus, AndroidReleaseTrack, SubmitProfile } from './submit/types';
export { getDefaultProfile as getDefaultSubmitProfile } from './submit/resolver';
export { EasJson, ProfileType } from './types';
export {
  AndroidVersionAutoIncrement,
  BuildProfile,
  CredentialsSource,
  DistributionType,
  IosEnterpriseProvisioning,
  IosVersionAutoIncrement,
} from './build/types';
export { EasJsonReader } from './reader';
export * as errors from './errors';
