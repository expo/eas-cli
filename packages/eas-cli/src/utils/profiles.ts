import { Platform } from '@expo/eas-build-job';
import {
  BuildProfile,
  EasJsonAccessor,
  EasJsonUtils,
  ProfileType,
  SubmitProfile,
} from '@expo/eas-json';
import fs from 'fs-extra';
import path from 'path';
import semver from 'semver';

import Log, { learnMore } from '../log';

type EasProfile<T extends ProfileType> = T extends 'build'
  ? BuildProfile<Platform>
  : SubmitProfile<Platform>;

export type ProfileData<T extends ProfileType> = {
  profile: EasProfile<T>;
  platform: Platform;
  profileName: string;
};

export async function getProfilesAsync<T extends ProfileType>({
  easJsonAccessor,
  platforms,
  profileName,
  type,
  projectDir,
}: {
  easJsonAccessor: EasJsonAccessor;
  platforms: Platform[];
  profileName?: string;
  projectDir: string;
  type: T;
}): Promise<ProfileData<T>[]> {
  const results = platforms.map(async function (platform) {
    const profile = await readProfileWithOverridesAsync({
      easJsonAccessor,
      platform,
      type,
      profileName,
      projectDir,
    });

    return {
      profile,
      profileName: profileName ?? 'production',
      platform,
    };
  });

  return await Promise.all(results);
}

async function setNodeVersionFromFileAsync(
  projectDir: string,
  profile: BuildProfile<Platform>
): Promise<void> {
  if (profile?.node) {
    return;
  }
  const nodeVersion = await getNodeVersionFromFileAsync(projectDir);
  if (nodeVersion) {
    Log.log(
      `The EAS build profile does not specify a Node.js version. Using the version specified in .nvmrc: ${nodeVersion} `
    );

    profile.node = nodeVersion;
  }
}

async function getNodeVersionFromFileAsync(projectDir: string): Promise<string | undefined> {
  const nvmrcPath = path.join(projectDir, '.nvmrc');
  if (!(await fs.pathExists(nvmrcPath))) {
    return;
  }

  let nodeVersion: string;
  try {
    nodeVersion = (await fs.readFile(nvmrcPath, 'utf8')).toString().trim();
  } catch {
    return undefined;
  }
  if (!semver.valid(nodeVersion)) {
    throw new Error(`Invalid node version in .nvmrc: ${nodeVersion}`);
  }
  return nodeVersion;
}

async function readProfileWithOverridesAsync<T extends ProfileType>({
  easJsonAccessor,
  platform,
  type,
  profileName,
  projectDir,
}: {
  easJsonAccessor: EasJsonAccessor;
  platform: Platform;
  type: T;
  profileName?: string;
  projectDir: string;
}): Promise<EasProfile<T>> {
  if (type === 'build') {
    const buildProfile = await EasJsonUtils.getBuildProfileAsync(
      easJsonAccessor,
      platform,
      profileName
    );

    await maybePrintBuildProfileDeprecationWarningsAsync(easJsonAccessor, platform, profileName);
    await setNodeVersionFromFileAsync(projectDir, buildProfile);

    return buildProfile as EasProfile<T>;
  } else {
    return (await EasJsonUtils.getSubmitProfileAsync(
      easJsonAccessor,
      platform,
      profileName
    )) as EasProfile<T>;
  }
}

let hasPrintedDeprecationWarnings = false;

/**
 * Only for testing purposes
 */
export function clearHasPrintedDeprecationWarnings(): void {
  hasPrintedDeprecationWarnings = false;
}

export async function maybePrintBuildProfileDeprecationWarningsAsync(
  easJsonAccessor: EasJsonAccessor,
  platform: Platform,
  profileName?: string
): Promise<void> {
  if (hasPrintedDeprecationWarnings) {
    return;
  }
  const deprecationWarnings = await EasJsonUtils.getBuildProfileDeprecationWarningsAsync(
    easJsonAccessor,
    platform,
    profileName ?? 'production'
  );
  if (deprecationWarnings.length === 0) {
    return;
  }
  Log.newLine();
  Log.warn('Detected deprecated fields in eas.json:');
  for (const warning of deprecationWarnings) {
    const warnlog: string = warning.message.map(line => `\t${line}`).join('\n');
    Log.warn(warnlog);
    if (warning.docsUrl) {
      Log.warn(`\t${learnMore(warning.docsUrl)}`);
    }
    Log.newLine();
  }
  hasPrintedDeprecationWarnings = true;
}
