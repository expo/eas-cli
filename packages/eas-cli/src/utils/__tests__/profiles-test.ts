import { Platform } from '@expo/eas-build-job';
import { EasJsonReader, errors } from '@expo/eas-json';

import { getProfilesAsync } from '../profiles';

jest.mock('@expo/eas-json', () => {
  const actual = jest.requireActual('@expo/eas-json');

  const EasJsonReaderMock = jest.fn();
  EasJsonReaderMock.prototype = {
    getBuildProfileAsync: jest.fn(),
    getSubmitProfileAsync: jest.fn(),
  };
  return {
    ...actual,
    EasJsonReader: EasJsonReaderMock,
  };
});

const getBuildProfileAsync = jest.spyOn(EasJsonReader.prototype, 'getBuildProfileAsync');

describe(getProfilesAsync, () => {
  afterEach(() => {
    getBuildProfileAsync.mockReset();
  });

  it('defaults to production profile', async () => {
    const result = await getProfilesAsync({
      projectDir: '/fake',
      platforms: [Platform.ANDROID, Platform.IOS],
      profileName: undefined,
      type: 'build',
    });

    expect(result[0].profileName).toBe('production');
    expect(result[1].profileName).toBe('production');
    expect(getBuildProfileAsync).toBeCalledWith(Platform.ANDROID, 'production');
    expect(getBuildProfileAsync).toBeCalledWith(Platform.IOS, 'production');
  });

  it('defaults to release profile when production profile is non-existent', async () => {
    getBuildProfileAsync
      .mockImplementationOnce(() => {
        throw new errors.MissingProfileError();
      })
      .mockImplementationOnce(() => {
        throw new errors.MissingProfileError();
      });

    const result = await getProfilesAsync({
      projectDir: '/fake',
      platforms: [Platform.ANDROID, Platform.IOS],
      profileName: undefined,
      type: 'build',
    });

    expect(result[0].profileName).toBe('release');
    expect(result[1].profileName).toBe('release');
    expect(getBuildProfileAsync).toBeCalledWith(Platform.ANDROID, 'production');
    expect(getBuildProfileAsync).toBeCalledWith(Platform.IOS, 'production');
    expect(getBuildProfileAsync).toBeCalledWith(Platform.ANDROID, 'release');
    expect(getBuildProfileAsync).toBeCalledWith(Platform.IOS, 'release');
  });

  it('fails when neither production or release profiles are present', async () => {
    getBuildProfileAsync.mockImplementation(() => {
      throw new errors.MissingProfileError();
    });

    await expect(
      getProfilesAsync({
        projectDir: '/fake',
        platforms: [Platform.ANDROID],
        profileName: undefined,
        type: 'build',
      })
    ).rejects.toThrowError(errors.MissingProfileError);
  });

  it('gets a specific profile', async () => {
    const result = await getProfilesAsync({
      projectDir: '/fake',
      platforms: [Platform.ANDROID, Platform.IOS],
      profileName: 'custom-profile',
      type: 'build',
    });

    expect(result[0].profileName).toBe('custom-profile');
    expect(result[1].profileName).toBe('custom-profile');
    expect(getBuildProfileAsync).toBeCalledWith(Platform.ANDROID, 'custom-profile');
    expect(getBuildProfileAsync).toBeCalledWith(Platform.IOS, 'custom-profile');
  });

  it('throws validation error if eas.json is invalid', async () => {
    getBuildProfileAsync.mockImplementation(() => {
      throw new errors.InvalidEasJsonError('eas.json is not valid');
    });

    await expect(
      getProfilesAsync({
        projectDir: '/fake',
        platforms: [Platform.ANDROID, Platform.IOS],
        profileName: undefined,
        type: 'build',
      })
    ).rejects.toThrowError(/eas.json is not valid/);
  });
});
