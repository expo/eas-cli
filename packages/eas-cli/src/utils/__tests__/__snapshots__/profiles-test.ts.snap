// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`maybePrintBuildProfileDeprecationWarningsAsync multiple deprecation warnings prints the warnings 1`] = `
[
  [
    "Detected deprecated fields in eas.json:",
  ],
  [
    "	The "build.production.cache.customPaths" field in eas.json is deprecated and will be removed in the future. Please use "build.production.cache.paths" instead.",
  ],
  [
    "	[2mLearn more: [4mhttps://docs.expo.dev/build-reference/eas-json/#cache[24m[22m",
  ],
  [
    "	The "build.production.cache.cacheDefaultPaths" field in eas.json is deprecated and will be removed in the future.",
  ],
  [
    "	[2mLearn more: [4mhttps://docs.expo.dev/build-reference/caching/#ios-dependencies[24m[22m",
  ],
  [
    "	Other message",
  ],
]
`;

exports[`maybePrintBuildProfileDeprecationWarningsAsync one deprecation warning prints the warning 1`] = `
[
  [
    "Detected deprecated fields in eas.json:",
  ],
  [
    "	The "build.production.cache.customPaths" field in eas.json is deprecated and will be removed in the future. Please use "build.production.cache.paths" instead.",
  ],
  [
    "	[2mLearn more: [4mhttps://docs.expo.dev/build-reference/eas-json/#cache[24m[22m",
  ],
]
`;
